querystring: check that maxKeys is finite

TheAlphaNerd opened the pull request:
There was a very subtle change in behavior introduced with 27def4f

In the past if querystring.parse was given Infinity for maxKeys,
everything worked as expected.

Check to see is maxKeys is Infinity before forwarding the value to
String.prototype.split which causes this regression

TheAlphaNerd commented:
/cc @jasnell @targos @thefourtheye

evanlucas commented:
LGTM, but should the test be in a separate commit?

TheAlphaNerd commented:
>  LGTM, but should the test be in a separate commit?

I'm really not sure, I can check previous commits. To me it makes sense to include the test with the commit removing the regression.

thefourtheye commented on lib/querystring.js:
```diff
@@ -209,20 +209,19 @@ QueryString.parse = QueryString.decode = function(qs, sep, eq, options) {
 
+  var len = qs.length;
   // maxKeys <= 0 means that we should not limit keys count
-  if (maxKeys > 0) {
```
Wouldn't it be better, if we just throw an `isFinite` check here?

TheAlphaNerd replied:
that is actually a great idea.

thefourtheye commented:
> The spec documents that the second argument must be an integer.
Infinity is being interpreted as 0... instead of MAX_INT
(perhaps this should be a fix upstream to v8).

Sadly, it is as per the spec only.

[`String.prototype.split`](http://es5.github.io/#x15.5.4.14) says,

> If limit is undefined, let lim = 232–1; else let lim = ToUint32(limit).

and [9.6 ToUint32: (Unsigned 32 Bit Integer)](http://es5.github.io/#x9.6) says,

> If number is NaN, +0, −0, +∞, or −∞, return +0.

TheAlphaNerd commented:
@thefourtheye great idea, I've restructured this commit

@evanlucas I'll need another LGTM. Rather than reverting I've added an extra check for infinity. It now makes much more sense to include the test

TheAlphaNerd commented:
@nodejs/testing is there a preference to writing more smaller tests, or putting this in with another test already testing querystring?

evanlucas commented:
Yea, I like this. LGTM. Definitely need CI on this one though to make sure arm is happy with it too

thefourtheye commented:
Adding a new test is fine in this case, and it would be better to have this issue also mentioned somewhere in the comments.

thefourtheye commented on test/parallel/test-querystring-maxKeys-Infinity.js:
```diff
@@ -0,0 +1,36 @@
+'use strict';
+const common = require('../common');
```
Didn't our linter complain about unused variable? I thought [this](https://github.com/nodejs/node/blob/master/.eslintrc#L105-L106),

```
  ## disallow declaration of variables that are not used in the code
  no-unused-vars: [2, {"args": "none"}]
```

forbids us

TheAlphaNerd replied:
yeah... I just pushed an update that fixes the linting issues

thefourtheye commented on test/parallel/test-querystring-maxKeys-Infinity.js:
```diff
@@ -0,0 +1,39 @@
+// 27def4f introduced a change to parse that would cause Inifity
+// to be passed to String.prototype.split as an argument for limit
+// In this instance split will always return an empty array
+// this test confirms that the output of parse is the expe4cted length
```
`expe4cted` -> `expected`

TheAlphaNerd commented on test/parallel/test-querystring-maxKeys-Infinity.js:
```diff
@@ -0,0 +1,39 @@
+const assert = require('assert');
+const parse = require('querystring').parse;
+
+// taken from express-js/body-parser
```
This code below was borrowed from
https://github.com/expressjs/body-parser/blob/master/test/urlencoded.js#L636-L651

I'm not sure what our policy is on this. FWIW body-parser is MIT

/cc @jasnell @rvagg @mikeal

thefourtheye commented on test/parallel/test-querystring-maxKeys-Infinity.js:
```diff
@@ -0,0 +1,39 @@
+// In this instance split will always return an empty array
+// this test confirms that the output of parse is the expe4cted length
+// when passed Infinity as the argument for maxKeys
+const result = parse(params, undefined, undefined, {maxKeys: Infinity});
```
Now that we are doing it, let's introduce tests with `NaN`, `'Infinity'` and `'NaN'`

TheAlphaNerd replied:
@thefourtheye I've updated the test to include tests for all four "non finite" values

TheAlphaNerd commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/1534/

thefourtheye commented:
I couldn't think of any other corner cases. LGTM.

cjihrig commented:
It might be a little more efficient to add the `isFinite()` check to the line `if (options && typeof options.maxKeys === 'number') {`. Then you would only get the overhead if a number is provided as input.

cjihrig commented:
LGTM with a comment.

thefourtheye commented:
In that case we can simply do `if (options && Number.isFinite(options.maxKeys)`

Trott commented:
> @nodejs/testing is there a preference to writing more smaller tests, or putting this in with another test already testing querystring?

Speaking for myself only, I definitely prefer smaller, isolated tests.

rvagg commented on test/parallel/test-querystring-maxKeys-non-finite.js:
```diff
@@ -0,0 +1,49 @@
+const assert = require('assert');
+const parse = require('querystring').parse;
+
+// taken from express-js/body-parser
```
put in a full github url including hash for the exact version you're pulling from

rvagg commented:
/cc @manvalls 

@TheAlphaNerd I think this is ok, lgtm



@trott:

> > @nodejs/testing is there a preference to writing more smaller tests, or putting this in with another test already testing querystring?

> Speaking for myself only, I definitely prefer smaller, isolated tests.

One thing to keep in mind is that separate tests take longer to run in CI than compacting them into fewer tests. Primarily due to very slow startup time on slower ARM machines but also it seems like this AIX and musl libc based systems have slow exit times (ref https://github.com/nodejs/node/pull/5056).

More files, separate tests is certainly the nicer way to approach it but I just wanted to register the above concern so that @nodejs/testing is aware of this and can (maybe) take it into account when forming opinions on how to best structure tests.

TheAlphaNerd commented:
@cjihrig @thefourtheye if we made the suggested change then maxKeys will always = 1000, this would also result in a behaviour change

TheAlphaNerd commented:
@rvagg I've updated the test with a comment that includes the github link, with sha and line numbers.

As you can expect this link is far over the max characters per line. I've split it on to two lines, hopefully that isn't confusing

cjihrig commented:
@TheAlphaNerd no, this is what I was proposing. It should be functionally equivalent to what you have, but slightly more efficient. And, as @thefourtheye pointed out, the last two conditions can be condensed to `Number.isFinite()`.

```javascript
  var maxKeys = 1000;
  if (options && typeof options.maxKeys === 'number' && isFinite(options.maxKeys)) {
    maxKeys = options.maxKeys;
  }

  // maxKeys <= 0 means that we should not limit keys count
  if (maxKeys > 0) {
```

TheAlphaNerd commented:
maxKeys is set just above though. Meaning that ```if(maxKeys > 0)``` will be true if ```options.MaxKeys === Infinity```

thus the split will be called with maxKeys set to 1000 which is still a behavior change.

To double check that I am correct I have implemented your change and run it against the tests and we get a failure

```AssertionError: 1000 == 10000```

cjihrig commented:
Sorry, you're right. I was thinking we wanted `maxKeys` to be the default value on bad input. The existing behavior seems silly... Use 1000 as a default, but if a number is provided use that, but if it's not positive and finite, ignore it completely.

thefourtheye commented:
> Use 1000 as a default, but if a number is provided use that, but if it's not positive and finite, ignore it completely.

Ya, that doesn't look right. Can we fix this behavior here?

TheAlphaNerd commented:
@thefourtheye I think we can fix that behavior, but that would be a Major change. In the mean time we do need to patch back the behavior that is expected (imho).

Or are you arguing that the behavior itself was a bug?

thefourtheye commented:
@TheAlphaNerd Though it seems odd, as you pointed out, our priority here is to get this fixed asap, with the current functionality. Perhaps we can improve this in a separate patch. LGTM.

TheAlphaNerd commented:
I think it is also worth mentioning that #5012 rewrite all the code this is touching.

ncopa commented:
> it seems like this AIX and musl libc based systems have slow exit times (ref #5056).

Technically, exit itself is quick. What happens is that the parent exits without reaping the child processes, so init (pid 1) has to reap those zombies. The slow timeout on Alpine Linux was due to busybox init has a "design flaw" that will intentionally delay the reaping of the orphaned child processes. This has actually nothing to do with musl libc and I am pretty sure that busybox init with glibc has the exact same issue. Most likely this is what AIX init does too.

Exit times are fast. But someone/something needs to reap the dead processes. Ideally should the parent process take care of that instead of letting init (pid 1) do it.

Fishrock123 commented:
Rubber-stamp LGTM

TheAlphaNerd commented:
Landed as 878bcd43f8

thefourtheye commented:
Express must be happy now, right?

rvagg commented:
Thanks @ncopa, that's very interesting background
